### ISSUE TITLE
AttributeError in custom layer for a weight defined in build() method when using JAX backend in Keras 3.0

### ISSUE THREAD

**Issue by Faptimus420** [keras-team/keras#18689]
I am reimplementing a TensorFlow implementation of a Conv2D layer with variational dropout in Keras Core. When I use this custom layer in a model with either the TF or the PyTorch backends, the network trains just fine. However, I get a `AttributeError: 'VariationalConv2d' object has no attribute 'theta'` error when I try to use the JAX backend.

In the custom layer, the theta attribute is defined using the add_weight function in the layer's build() method:
```python
def build(self, input_shape):
    self.theta = self.add_weight(name="kernel", shape=self.kernel_size, initializer=self.kernel_initializer, trainable=True, regularizer=self.kernel_regularizer)
```
in my testing, I use (5, 5, 1, 6) as self.kernel_size, 'glorot_normal' as self.kernel_initializer, and None as self.kernel_regularizer.
    
In addition to the standard methods, this property is also defined within the layer:
```python
@property
    def log_alpha(self):
        """Compute log alpha, which is the log of the variance of the weight matrix."""
        theta = ops.where(ops.isnan(self.theta), ops.zeros_like(self.theta), self.theta)
        log_sigma2 = ops.where(ops.isnan(self.log_sigma2), ops.zeros_like(self.log_sigma2), self.log_sigma2)
        log_alpha = ops.clip(log_sigma2 - ops.log(ops.square(theta) + 1e-10), -20.0, 4.0)
        return ops.where(ops.isnan(log_alpha), self.threshold * ops.ones_like(log_alpha), log_alpha)
```
        
The above property gets called by the following regularization property, whose result is added onto the result of the loss function:
```python
@property
    def regularization(self):
        """Compute the regularization term for the weight matrix."""
        k1, k2, k3 = 0.63576, 1.8732, 1.48695
        C = -k1
        mdkl = k1 * ops.sigmoid(k2 + k3 * self.log_alpha) - 0.5 * ops.log(1 + (ops.exp(-self.log_alpha))) + C
        return -ops.sum(mdkl)
```
        
According to the traceback of the AttributeError exception, the call causing the exception is the line `theta = ops.where(ops.isnan(self.theta), ops.zeros_like(self.theta), self.theta)` within log_alpha.

For clarity, I will add the full layer definition and the train_step and training loop definitions in a comment.

**Comment by Faptimus420** [comment]
Full layer definition:
```python
class VariationalConv2d(Layer):
    def __init__(self, kernel_size, stride, padding='SAME', threshold=3.0, activation=None, data_format='channels_last', kernel_initializer='glorot_normal', bias_initializer='zeros', kernel_regularizer=None):
        super(VariationalConv2d, self).__init__()
        assert len(kernel_size) == 4    #kernel_size: [filter_height, filter_width, in_channels, out_channels]
        self.kernel_size = kernel_size
        self.stride = stride
        self.padding = padding
        self.threshold = threshold
        self.activation = activations.get(activation)
        self.data_format = data_format

        self.kernel_initializer = initializers.get(kernel_initializer)
        self.bias_initializer = initializers.get(bias_initializer)

        self.kernel_regularizer = regularizers.get(kernel_regularizer)

    def build(self, input_shape):
        self.theta = self.add_weight(name="kernel", shape=self.kernel_size,
                                     initializer=self.kernel_initializer, trainable=True, regularizer=self.kernel_regularizer)

        self.log_sigma2 = self.add_weight(name="log_sigma2", shape=self.kernel_size,
                                 initializer=initializers.Constant(-10.0), trainable=True)

    def sparsity(self):
        """Compute sparsity of the weight matrix, based on the number of non-zero elements."""
        total_param = ops.prod(ops.shape(self.boolean_mask))
        remaining_param = ops.count_nonzero(ops.cast(self.boolean_mask, dtype="uint8"))

        return remaining_param, total_param

    @property
    def log_alpha(self):
        """Compute log alpha, which is the log of the variance of the weight matrix."""
        theta = ops.where(ops.isnan(self.theta), ops.zeros_like(self.theta), self.theta)
        log_sigma2 = ops.where(ops.isnan(self.log_sigma2), ops.zeros_like(self.log_sigma2), self.log_sigma2)
        log_alpha = ops.clip(log_sigma2 - ops.log(ops.square(theta) + 1e-10), -20.0, 4.0)
        return ops.where(ops.isnan(log_alpha), self.threshold * ops.ones_like(log_alpha), log_alpha)

    @property
    def boolean_mask(self):
        return self.log_alpha <= self.threshold

    @property
    def sparse_theta(self):
        theta = ops.where(ops.isnan(self.theta), ops.zeros_like(self.theta), self.theta)
        return ops.where(self.boolean_mask, theta, ops.zeros_like(theta))

    @property
    def regularization(self):
        """Compute the regularization term for the weight matrix."""
        k1, k2, k3 = 0.63576, 1.8732, 1.48695
        C = -k1
        mdkl = k1 * ops.sigmoid(k2 + k3 * self.log_alpha) - 0.5 * ops.log(1 + (ops.exp(-self.log_alpha))) + C

        return -ops.sum(mdkl)


    def call(self, input, **kwargs):
        """Forward pass of the layer - differentiates between sparse (on evaluation time) and dense (on training time) input."""
        theta = ops.where(ops.isnan(self.theta), ops.zeros_like(self.theta), self.theta)

        if not kwargs['sparse_input']:
            sigma = ops.sqrt(ops.exp(self.log_alpha) * theta * theta)
            self.weight = theta + random.normal(ops.shape(theta), 0.0, 1.0) * sigma
            output = ops.conv(input, self.weight, self.stride, self.padding, self.data_format)

            if self.activation is not None:
                output = self.activation(output)

        else:
            output = ops.conv(input, self.sparse_theta, self.stride, self.padding, self.data_format)
            if self.activation is not None:
                output = self.activation(output)

        return output


    def get_config(self):
        config = super().get_config()
        config.update(
            {
                "kernel_size": self.kernel_size,
                "strides": self.strides,
                "padding": self.padding,
                "data_format": self.data_format,
                "activation": activations.serialize(self.activation),
                "kernel_initializer": initializers.serialize(
                    self.kernel_initializer
                ),
                "bias_initializer": initializers.serialize(
                    self.bias_initializer
                ),
                "kernel_regularizer": regularizers.serialize(
                    self.kernel_regularizer
                ),
                "threshold": self.threshold,
            }
        )
        return config
```
        
JAX training loop:
```python
        def compute_loss(t, preds, reg):
            return criterion(t, preds) + reg

        def compute_loss2(label, pred):
            return criterion(label, pred)

        def compute_loss_and_updates(trainable_variables, non_trainable_variables, metric_variables, x, t, epoch,
                                     training=False):
            reg = rw_schedule(epoch) * model.regularization()

            preds, non_trainable_variables = model.stateless_call(trainable_variables, non_trainable_variables, x,
                                                                  train=training, sparse_input=False)

            loss = compute_loss(t, preds, reg)
            metric_variables = train_acc.stateless_update_state(metric_variables, t, preds)
            return loss, (non_trainable_variables, metric_variables)

        grad_fn = jax.value_and_grad(compute_loss_and_updates, has_aux=True)

        @partial(jax.jit, static_argnames=['epoch'])
        def train_step(state, data, epoch):
            (
                trainable_variables,
                non_trainable_variables,
                optimizer_variables,
                metric_variables,
            ) = state
            x, t = data

            (loss, (non_trainable_variables, metric_variables)), grads = grad_fn(trainable_variables,
                                                                                 non_trainable_variables,
                                                                                 metric_variables,
                                                                                 x, t, epoch, True)

            trainable_variables, optimizer_variables = optimizer.stateless_apply(optimizer_variables, grads,
                                                                                 trainable_variables)

            # Return updated state
            return loss, (
                trainable_variables,
                non_trainable_variables,
                optimizer_variables,
                metric_variables,
            )

        @jax.jit
        def eval_step(state, data):
            trainable_variables, non_trainable_variables, metric_variables = state
            x, t = data

            preds, non_trainable_variables = model.stateless_call(trainable_variables, non_trainable_variables,
                                                                  x, train=False, sparse_input=True)

            loss = compute_loss2(t, preds)

            metric_variables = test_acc.stateless_update_state(metric_variables, t, preds)

            return loss, (
                trainable_variables,
                non_trainable_variables,
                metric_variables,
            )


        # Initialization
        optimizer.build(model.trainable_variables)
        trainable_variables = model.trainable_variables
        non_trainable_variables = model.non_trainable_variables
        optimizer_variables = optimizer.variables
        metric_variables = train_acc.variables
        state = (
            trainable_variables,
            non_trainable_variables,
            optimizer_variables,
            metric_variables,
        )

        for epoch in range(epochs):
            x_train, y_train = shuffle(x_train, y_train, random_state=42)

            for batch in range(n_batches):
                start = batch * batch_size
                end = start + batch_size
                x_batch, y_batch = x_train[start:end], y_train[start:end]
                loss, state = train_step(state,
                                         (ops.convert_to_tensor(x_batch, dtype="float32"),
                                          ops.convert_to_tensor(y_batch, dtype="float32")),
                                         epoch)

            loss, state = eval_step(state,
                                    (ops.convert_to_tensor(x_test, dtype="float32"),
                                     ops.convert_to_tensor(y_test, dtype="float32")))
            for variable, value in zip(train_acc.variables, metric_variables):
                variable.assign(value)

            print(f'Epoch: {epoch + 1}, Valid Cost: {float(loss):.3f}, Valid Acc: {test_acc.result():.3f}')
            print("Sparsity: ", ops.convert_to_numpy(model.count_sparsity()))
```

Model definition:
```python
def rw_schedule(epoch):
    """Defines the schedule for the weight regularization term."""
    if epoch <= 1:
        return 0
    else:
        return 0.0001 * (epoch - 1)


class VariationalLeNet(Model):
    def __init__(self, n_class=10):
        super().__init__()
        self.n_class = n_class

        self.conv1 = VariationalConv2d((5, 5, 1, 6), stride=1, padding='VALID')
        self.pooling1 = MaxPooling2D(padding='SAME')
        self.conv2 = VariationalConv2d((5, 5, 6, 16), stride=1, padding='VALID')
        self.pooling2 = MaxPooling2D(padding='SAME')

        self.flat = Flatten()
        self.fc1 = VariationalDense(120)
        self.fc2 = VariationalDense(84)
        self.fc3 = VariationalDense(10)

        self.hidden_layers = [self.conv1, self.conv2, self.fc1, self.fc2, self.fc3]

    def call(self, input, **kwargs):
        x = self.conv1(input, sparse_input=kwargs['sparse_input'])
        x = ops.relu(x)
        x = self.pooling1(x)
        x = self.conv2(x, sparse_input=kwargs['sparse_input'])
        x = ops.relu(x)
        x = self.pooling2(x)
        x = self.flat(x)
        x = self.fc1(x, sparse_input=kwargs['sparse_input'])
        x = ops.relu(x)
        x = self.fc2(x, sparse_input=kwargs['sparse_input'])
        x = ops.relu(x)
        x = self.fc3(x, sparse_input=kwargs['sparse_input'])
        return ops.softmax(x)

    def regularization(self):
        """Computes the total regularization term that has been applied on all the layers."""
        total_reg = 0
        for layer in self.hidden_layers:
            total_reg += layer.regularization

        return total_reg

    def count_sparsity(self):
        """Computes the sparsity of the weight matrices in all the layers."""
        total_remain, total_param = 0, 0
        for layer in self.hidden_layers:
            a, b = layer.sparsity()
            total_remain += a
            total_param += b

        return 1 - (total_remain / total_param)
```

**Comment by haifeng-jin** [comment]
@Faptimus420  Would you mind creating a colab for reproducing the error?
So that we can look further into it.

Thanks!

**Comment by Faptimus420** [comment]
@haifeng-jin Sure, here it is https://colab.research.google.com/drive/1Oo2RZ00T3dbOCCRa68kL6hfk7dqu4LVc

**Comment by fchollet** [comment]
I looked into this a bit.

JAX is much stricter than other backends about creating state in advance of using the model (due to its stateless nature -- it cannot create state on the fly like the other backends). In particular, **you cannot call an unbuilt model in a JAX jit scope**. The state must have been created beforehand.

So the first thing is that you need to build the model before using it. You can do so e.g. before calling `optimizer.build()`. Just add `model.build(input_shape)`, e.g. `model.build(x_train.shape[0:1])`.

When you do this you will see an error about your custom layers having unbuilt state and needing to build their sublayers. This is because calling `build()` on the model will create the sublayers but won't build their variables. So you will need to add a `build()` method on your `VariationalLeNet` class that builds the sublayers. It could look like this:

```python
    def build(self, input_shape):
        for layer in self.hidden_layers:
            layer.build(input_shape)
            input_shape = layer.compute_output_shape(input_shape)
```

In order to implement it, you will further need to add `def compute_output_shape(self, input_shape)` on your two custom layers, `VariationalConv2d` and `VariationalDense`.

So in summary, you should add:

- `VariationalDense.compute_output_shape`
- `VariationalConv2d.compute_output_shape`
- `VariationalLeNet.build`

And then you should call `model.build(input_shape)` before using the model (outside the jit scope).

**Comment by fchollet** [comment]
Oh and quick note -- if you do not want to bother with `compute_output_shape` should can build the model via something like:

```python
def build(self, input_shape):
   # eager build
   x = keras.ops.zeros(input_shape)
   self(x)
```

In your case it would be a tad more complex because your model takes more than one call argument.
